bpo-35193: Fix an off by one error in the RETURN_VALUE case.

gpshead opened the pull request:
By doing a bounds check within find_op so it can
return before going past the end as a safety measure.

https://github.com/python/cpython/commit/7db3c488335168993689ddae5914a28e16188447#diff-a33329ae6ae0bb295d742f0caf93c137
introduced this off by one error while fixing another one nearby.



<!-- issue-number: [bpo-35193](https://bugs.python.org/issue35193) -->
https://bugs.python.org/issue35193
<!-- /issue-number -->

gpshead committed:
Fix an off by one error in the RETURN_VALUE case.

By doing a bounds check within find_op so it can
return before going past the end as a safety measure.

https://github.com/python/cpython/commit/7db3c488335168993689ddae5914a28e16188447#diff-a33329ae6ae0bb295d742f0caf93c137
introduced this off by one error while fixing another one nearby.

serhiy-storchaka commented on Python/peephole.c:
```diff
@@ -417,7 +419,7 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
                 }
                 if (h > i + 1) {
                     fill_nops(codestr, i + 1, h);
-                    nexti = find_op(codestr, h);
```
Wouldn't be simpler to just add a condition in this line?
```c
nexti = (h < codelen) ? find_op(codestr, h) : codelen;
```

Yhg1s replied:
That would leave room for the same bug to be introduced in other calls to find_op.

serhiy-storchaka replied:
In other calls `find_op()` can't be called with `codelen`.

Yhg1s replied:
Yes. That's why I said "introduced". this prevents the bug from reappearing when the rest of the code changes.

gpshead committed:
Add a NEWS entry.

gpshead committed:
Reword the NEWS entry.

gpshead committed:
Add a unittest that will fail under clang msan.

I don't see how to test for the presence of this out of bounds memory
access issue itself from the level of the CPython interpreter, but this
test will reliably fail (crash) when the interpreter is built using the
clang memory sanitizer.

gpshead committed:
Mention the releases that this bug is in.

Yhg1s commented on Misc/NEWS.d/next/Core and Builtins/2018-11-08-15-00-58.bpo-35193.HzPS6R.rst:
```diff
@@ -0,0 +1,5 @@
+Fix an off by one error in the bytecode peephole optimizer where it could read
+bytes beyond the end of bounds of a bytecode array when removing unreachable
+code after a return.  This did not always crash an interpreter, but could.
+An interpreter compiled using the clang memory sanitizer always failed.
```
This seems a little longer and more detailed than most news entries... not sure if that's good or bad.

gpshead committed:
Shorten the news entry text.
